### PR TITLE
Refactor harmonizer

### PR DESCRIPTION
**Pull request by kurtwheeler**
## Issue Number

#2828 

## Purpose/Implementation Notes

Try to make the harmonizer more reliable, readable, and maintainable. Also try to address the issue https://github.com/AlexsLemonade/refinebio/pull/2807 was trying to fix.

I was able to change the GEO and SRA surveyors to harmonize the metadata sample by sample, which should eliminate the kind of mismatch errors that could exist before. Array express can't be done that way because we're combining two different responses, metadata from one and file data from the other.

## Types of changes

- refactor, bug fixes

## Functional tests

Unit tests are passing! I added one for the experiment that wouldn't survey before 

## Checklist

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: foreman/data_refinery_foreman/surveyor/harmony.py [hunk 1]
```diff
@@ -697,33 +357,303 @@ def parse_sdrf(sdrf_url: str) -> List:
...
+            "strain/genotype",
+        ]
+        self.genetic_information_fields = create_variants(genetic_information_fields)
+
+        disease_fields = [
+            "disease",
+            "disease state",
+            "disease status",
+            "diagnosis",
+            "disease",
```
Comment (wvauclain) [inline foreman/data_refinery_foreman/surveyor/harmony.py:475]: Is this redundant?

File: foreman/data_refinery_foreman/surveyor/test_harmony.py [hunk 2]
```diff
@@ -271,48 +280,67 @@ def test_geo_harmony(self):
         """
         Thoroughly tests a specific GEO harmonization
         """
-
         # Weird ones caused bugs
         gse = GEOparse.get_GEO("GSE94532", destdir="/tmp/GSE94532/", silent=True)
-        preprocessed_samples = preprocess_geo(gse.gsms.items())
+        for _, sample in gse.gsms.items():
```
Comment (wvauclain) [inline foreman/data_refinery_foreman/surveyor/test_harmony.py:285]: What are we checking for here?

**Review by wvauclain — COMMENTED** [review]
Almost all of the PR looks really good, I just have one question about the sample title changes

File: foreman/data_refinery_foreman/surveyor/test_harmony.py (hunk 2 above)
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/test_harmony.py:285]: From the comment I assume that calling preproces_geo() used to cause errors on this experiment until it got fixed?

It'd be nice if the comment was a little bit more descriptive... I'll add a couple checks of some sort here.

File: foreman/data_refinery_foreman/surveyor/harmony.py [hunk 3]
```diff
@@ -10,657 +254,73 @@
 logger = get_and_configure_logger(__name__)
 
 
-def extract_title(sample: Dict) -> str:
+TITLE_FIELDS = [
```
[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/harmony.py:257]: Hmm, I'm not sure that I needed to but I think this order makes the most sense... While I'm not sure this change is necessary, I think it's probably worth doing. I can't help but think that if a sample has both a `sample name` and a `labeled extract name` that the `sample name` field is more likely to cause us to succeed at processing if we're using it as a column name.

[the remark this answers is not in the document]
Comment (kurtwheeler) [reply foreman/data_refinery_foreman/surveyor/harmony.py:257]: That's certainly an assumption so I'll run the surveyor for some or all of the experiments in our processor tests and see if there's any differences.

**Comment by kurtwheeler** [comment]
@wvauclain I've rerun the surveyor for the two experiments in `test_illumina`. Here are the sample titles, which match exactly what's already in the tests:
```
data_refinery=# select title from samples order by created_at desc;
       title        
--------------------
 CB CD34+ normoxia
 CB CD34+ hypoxia
 LV-T350A&si-EZH2-3
 LV-T350A&si-EZH2-2
 LV-T350A&si-EZH2-1
 LV-EZH2&si-EZH2-3
 LV-EZH2&si-EZH2-2
 LV-EZH2&si-EZH2-1
 LV-C&si-EZH2-3
 LV-C&si-EZH2-2
 LV-C&si-EZH2-1
 LV-C&si-Control-3
 LV-C&si-Control-2
 LV-C&si-Control-1
```

I didn't see any other processors that used the `title` field. Were there any others that you were aware of?

**Review by wvauclain — APPROVED** [review]
None that I'm aware of, so we should be good to go!

**Pull request merged**